v3/src/components/graph/components/legend/numeric-legend.tsx

```diff
@@ -1,18 +1,93 @@
+import {measureTextExtent} from "../../../../hooks/use-measure-text"
+import {useDataSetContext} from "../../../../hooks/use-data-set-context"
+import {kChoroplethHeight, kGraphFont} from "../../graphing-types"
+import {axisGap} from "../../../axis/axis-types"
 
 interface INumericLegendProps {
   legendAttrID: string
-  transform:string
+  transform: string
 }
 
-export const NumericLegend = memo(function NumericLegend({legendAttrID}: INumericLegendProps) {
-/*
+export const NumericLegend = memo(function NumericLegend({transform, legendAttrID}: INumericLegendProps) {
   const dataConfiguration = useDataConfigurationContext(),
-    values = dataConfiguration?.numericValuesForAttrRole('legend')
-*/
+    layout = useGraphLayoutContext(),
+    dataset = useDataSetContext(),
+    quantileScale = useRef<ScaleQuantile<string>>(scaleQuantile()),
+    [choroplethElt, setChoroplethElt] = useState<SVGGElement | null>(null),
+    valuesRef = useRef<number[]>([]),
 
-  return (
-    <></>
-  )
+    computeDesiredExtent = useCallback(() => {
+      const labelHeight = measureTextExtent(dataset?.attrFromID(legendAttrID).name ?? '', kGraphFont).height
+      return 2 * labelHeight + kChoroplethHeight + 2 * axisGap
+    }, [dataset, legendAttrID]),
+
+    refreshScale = useCallback(() => {
+      if (choroplethElt) {
+        valuesRef.current = dataConfiguration?.numericValuesForAttrRole('legend') ?? []
+        layout.setDesiredExtent('legend', computeDesiredExtent())
+        quantileScale.current.domain(valuesRef.current).range(schemeBlues[5])
+        const bounds = layout.computedBounds.get('legend'),
+          translate = `translate(${bounds?.left}, ${(bounds?.top ?? 0) + axisGap})`
+        choroplethLegend(quantileScale.current, choroplethElt,
+          {
+            transform: translate, width: bounds?.width,
+            marginLeft: 6, marginRight: 6, ticks: 5,
+            clickHandler: (quantile: number, extend: boolean) => {
+              dataConfiguration?.selectCasesForLegendQuantile(quantile, extend)
+            },
+            casesInQuantileSelectedHandler: (quantile: number) => {
+              return !!dataConfiguration?.casesInQuantileSelected(quantile)
+            }
+          })
+      }
+    }, [layout, computeDesiredExtent, choroplethElt, dataConfiguration])
+
+  useEffect(function refresh() {
+    refreshScale()
+  }, [refreshScale])
+
+  useEffect(function respondToLayoutChange() {
+    const disposer = reaction(
+      () => {
+        const {graphHeight, graphWidth} = layout,
+          legendID = dataConfiguration?.attributeID('legend')
+        return [graphHeight, graphWidth, legendID]
+      },
+      () => {
+        refreshScale()
+      }, {fireImmediately: true}
+    )
+    return () => disposer()
+  }, [layout, computeDesiredExtent, dataConfiguration, refreshScale])
+
+  useEffect(function respondToSelectionChange() {
+    const disposer = onAction(dataset, action => {
```

Comment:
It shouldn't make any difference functionally at this point, but I think it would make more sense to restrict your reactions to the `DataConfigurationModel`:
```suggestion
    const disposer = dataConfiguration.onAction(action => {
```
We could conceivably filter out notifications for cases that don't pass the filter, for instance.